Datastore/Sources/Data API/API/ProjectAPI.swift

```diff
@@ -0,0 +1,176 @@
+    func rollback(transactionId: String) -> EventLoopFuture<EmptyResponse>
+    
+    /// Queries for entities.
+    /// - Parameters:
+    ///   - partitionId: Entities are partitioned into subsets, identified by a partition ID. Queries are scoped to a single partition. This partition ID is normalized with the standard default context partition ID.
+    ///   - readOptions: The options for this query.
+    ///   - datastoreQuery: A query to run, either of normal or GQL type
+    func runQuery(partitionId: PartitionId, readOptions: ReadOptions?, datastoreQuery: DatastoreQuery) -> EventLoopFuture<RunQueryResponse>
+}
+
+extension DatastoreProjectAPI {
+    
+    public func allocateIDs(keys: [Key]) -> EventLoopFuture<AllocateIdsResponse> {
+        return allocateIDs(keys: keys)
+    }
+    
+    func reserveIDs(databaseId: String? = nil, keys: [Key]) -> EventLoopFuture<EmptyResponse> {
+        return reserveIDs(databaseId: databaseId, keys: keys)
+    }
+    
+    public func beginTransaction(transactionOptions: TransactionOptions) -> EventLoopFuture<BeginTransactionResponse> {
+        return beginTransaction(transactionOptions: transactionOptions)
+    }
+    
+    public func commit(mode: CommitRequest.Mode = .transactional, mutations: [CommitRequest.Mutation], transactionId: String) -> EventLoopFuture<CommitResponse> {
+        return commit(mode: mode, mutations: mutations, transactionId: transactionId)
+    }
+    
+    public func lookup(keys: [Key]) -> EventLoopFuture<LookupResponse> {
+        return lookup(keys: keys)
+    }
+    
+    public func rollback(transactionId: String) -> EventLoopFuture<EmptyResponse> {
+        return rollback(transactionId: transactionId)
+    }
+   
+    public func runQuery(partitionId: PartitionId, readOptions: ReadOptions? = nil, datastoreQuery: DatastoreQuery) -> EventLoopFuture<RunQueryResponse> {
+        return runQuery(partitionId: partitionId, readOptions: readOptions, datastoreQuery: datastoreQuery)
+    }
+}
+
+public final class GoogleCloudDatastoreProjectAPI: DatastoreProjectAPI {
+    
+    let endpoint: String
+    let request: GoogleCloudDatastoreRequest
+    let encoder = JSONEncoder()
+    
+    init(request: GoogleCloudDatastoreRequest,
+         endpoint: String) {
+        self.request = request
+        self.endpoint = endpoint
+    }
+    
+    private var projectPath: String {
+        "\(endpoint)/v1/projects/\(request.project)"
+    }
+    
+    public func allocateIDs(keys: [Key]) -> EventLoopFuture<AllocateIdsResponse> {
+        
+        do {
+            let allocateIdsRequest = AllocateIdsRequest(keys: keys)
+            let body = try HTTPClient.Body.data(encoder.encode(allocateIdsRequest))
+            return request.send(method: .POST, path: "\(projectPath):allocateIds", body: body)
+        } catch {
+            return request.httpClient.eventLoopGroup.next().makeFailedFuture(error)
```

Comment:
Also we can actually just use `request.eventLoop` for all the catch blocks in the requests and not have to ask the client anymore. Some of this in the cloud storage API was leftover and I'll raise a PR to clean it up soon.